File: HW1/323131268_111111111_assignment1/323131268_111111111_assignment1/kmeans.py
# Import necessary modules
import copy
import sys


def main( ):
    # initialize the variables
    try:
        if len(sys.argv) == 5:
            k = int(sys.argv[1])
            n = int(sys.argv[2])
            dimensions = int(sys.argv[3])
            max_iterations = 200
            input_file = sys.argv[4]
        elif len(sys.argv) == 6:
            k = int(sys.argv[1])
            n = int(sys.argv[2])
            dimensions = int(sys.argv[3])
            max_iterations = int(sys.argv[4])
            input_file = sys.argv[5]

        else:
            print("An Error Has Occured")
            return
        if k < 1:
            print("Invalid number of clusters!")
            return
        if n < 1 or n < k:
            print("Invalid number of points!")
            return
        if dimensions < 1 :
            print("Invalid number of dimensions!")
            return
        if max_iterations < 1 or max_iterations > 1000:
            print("Invalid number of iterations!")
            return
    
    
        centeroids = [] #matrix of centeroids
        points = [] #matrix of points

        count = 0
        with open(input_file, 'r') as file:
            for line in file:
                #initize  the points array, the final index in each row is the cluster index 
                points.append([float(x) for x in line.strip().split(',')] + [0])
                if count < k:
                    centeroids.append([float(x) for x in line.strip().split(',')])
                    count += 1

        #main algohrithm
        kMeans(k,centeroids,points,dimensions,max_iterations)
    
    except Exception as e:
        print("An Error Has Occured", e)



    return
    


def kMeans(k,centeroids, points, dimensions,max_iterations = 200):
    # Implement the kMeans algorithm
    tmpCenteroids = copy.deepcopy(centeroids)
    pointsInCluster = [0 for i in range(k)]
    
    epsilon = 0.001
    changes = 1.000 
    for i in range(max_iterations):
        #stop iterating if the changes are less than epsilon
        if changes <= epsilon:
            break

        #initize the temp centeroid holder to zero
        for i in range(len(tmpCenteroids)):
            for j in range(len(tmpCenteroids[i])):
                tmpCenteroids[i][j] = 0
            pointsInCluster[i] = 0
        
        for point in points:
            minDistance = calculateDistance(point,centeroids[0],dimensions)
            minCluster = 0
            #calculate the distance between the point and each of the centeroids and get the minimum distance
            for j in range(k):
                tmpDistance = calculateDistance(point,centeroids[j],dimensions)
                if tmpDistance < minDistance:
                    minDistance = tmpDistance
                    minCluster = j
            #update the cluster index of the point and calculate the changes from past cluster:
            # if the point is in a new cluster, add the point to the new cluster and remove it from the old cluster
            if(point[dimensions] != minCluster):
                #point[dimensions] is the cluster index of the point
                point[dimensions] = minCluster
                for t in range(dimensions):
                    tmpCenteroids[minCluster][t] = tmpCenteroids[minCluster][t] + point[t]
                pointsInCluster[minCluster] = pointsInCluster[minCluster] + 1
            #if the point is in the same cluster, add the point to the cluster that is saved in the point last index (dimensions)
            else:
                for t in range(dimensions):
                    tmpCenteroids[point[dimensions]][t] = tmpCenteroids[point[dimensions]][t] + point[t]
                pointsInCluster[point[dimensions]] = pointsInCluster[point[dimensions]] + 1    

        #update the centeroids
        changes = 0
        for i in range(k):
            if pointsInCluster[i] == 0:
                #for j in range(dimensions):
                #tmpCenteroids[i][j] = centeroids[i][j] 
                continue
            else:
                for j in range(dimensions):
                    tmpCenteroids[i][j] = tmpCenteroids[i][j] / pointsInCluster[i]

        #calculate the changes from the past centeroids and update the centeroids from the temp centeroids to the real centeroids
        for i in range(k):
            changes = changes + calculateDistance(tmpCenteroids[i],centeroids[i],dimensions)
            for j in range(dimensions):
                centeroids[i][j] = tmpCenteroids[i][j]

    #print centeroids:
    for i in range(k):
        for j in range (dimensions-1):
            print(round(centeroids[i][j],4), end = ",")
        print(round(centeroids[i][dimensions-1],4))

    return

    

def calculateDistance(point1, point2,dimensions):
    # Implement the distance calculation between two points
    calcualted_distance = 0
    for i in range(dimensions):
        calcualted_distance += (point1[i] - point2[i])**2
    calcualted_distance = calcualted_distance**0.5
    
    return calcualted_distance




if __name__ == "__main__":
    main()

File: HW1/323131268_111111111_assignment1/kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double calculateDistance(double point1[], double point2[], int dimensions){
    double distance = 0;
    int i;
   

    for(i = 0; i < dimensions; i++){
        distance += (point1[i] - point2[i]) * (point1[i] - point2[i]);
    }
    distance = sqrt(distance);
    
    
    return distance;
} 


int main(int argc, char *argv[]) {
    int i,j,t;
    double **clusterMatrix;
    double **pointMatrix;
    double **tempClusters; 
    int *pointsInCluster;
    FILE *file;

    
    double epsilon;
    double change;
    int minCluster;
    double minDistance;
    double distance;




    int k; 
    int n;
    int d;
    int max_iter;
    if (argc == 4){
        k = atoi(argv[1]);
        n = atoi(argv[2]);
        d = atoi(argv[3]);
        max_iter = 200;
    }
    else if (argc == 5){
        k = atoi(argv[1]);
        n = atoi(argv[2]);
        d = atoi(argv[3]);
        max_iter = atoi(argv[4]);
    }
    else{
        printf("An Error Has Occured\n");
        return 1;
    }
    /*initailize the matrix for clusters and points and a temp matrix for clusters*/
    clusterMatrix = (double **)malloc(k * sizeof(double *));
    for (i = 0; i < k; i++){
        clusterMatrix[i] = (double *)malloc(d * sizeof(double));
    }
    pointMatrix = (double **)malloc(n * sizeof(double *));
    for (i = 0; i < n; i++){
        pointMatrix[i] = (double *)malloc((d+1)* sizeof(double));
    }
    tempClusters = (double **)malloc(k * sizeof(double *));
    for (i = 0; i < k; i++){
        tempClusters[i] = (double *)malloc(d * sizeof(double));
    }
    /* initialize the pointsInCluster array stating how many points are in each cluster */
    pointsInCluster = (int *)malloc(k * sizeof(int));
    for (i = 0; i < k; i++){
        pointsInCluster[i] = 0;
    }

    

    file = stdin; 
    if (file == NULL){
        printf("An Error Has Occured\n");
        return 1;
    }

    
    for (i = 0; i < n; i++) {
        for (j = 0; j <= d; j++) {
            /* The final index in each pointer is the cluster index so we set it to 0 */
            if (j == d){
                pointMatrix[i][j] = 0;
            }
            /* reads into pointerMatrix from file */
            else if (fscanf(file, "%lf,", &pointMatrix[i][j]) != 1) {
                printf("An Error Has Occured\n");
                printf("Failed to read number at row %d, column %d\n", i, j);
                return 1;
            }
            else if (i < k){
                /* reads first k values into clusterMatrix */
                clusterMatrix[i][j] = pointMatrix[i][j];
            }
        }
    }
    fclose(file);
    


    epsilon = 0.001;
    change = 1.000;
    minCluster = 0;
    minDistance = 0;
    distance = 0;

    for (t = 0; t < max_iter; t++){
        /* if the change from last iteration is less than epsilon, break */
        if (change < epsilon){
            break;
        }
        /* initialize tempClusters to zero
        initialize pointsInCluster to zero */
        for ( i = 0; i < k; i++){
            for ( j = 0; j < d; j++){
                tempClusters[i][j] = 0;
            }
            pointsInCluster[i] = 0;
        }
        /* assign each point to the closest cluster based on distance */
        for ( i = 0; i < n; i++){
            minDistance = calculateDistance(pointMatrix[i], clusterMatrix[0], d);
            minCluster = 0;
            for (j = 1; j < k; j++){
                distance = calculateDistance(pointMatrix[i], clusterMatrix[j], d);
                if (distance < minDistance){
                    minDistance = distance;
                    minCluster = j;
                }
            }
            if(pointMatrix[i][d] != minCluster){
                pointMatrix[i][d] = minCluster;
                pointsInCluster[minCluster]++;
                for (j = 0; j < d; j++){
                    tempClusters[minCluster][j] += pointMatrix[i][j];
                }
            }
            else {
                pointsInCluster[minCluster]++;
                for (j = 0; j < d; j++){
                    tempClusters[minCluster][j] += pointMatrix[i][j];
                }
            }
        }
        
        /*calculate the new clusters*/
        for ( i = 0; i < k; i++){
            for ( j = 0; j < d; j++){
                tempClusters[i][j] = tempClusters[i][j] / pointsInCluster[i];
            }
        }
        /*check if the clusters have changed and add change to the total change
        set the new clusters to the clusterMatrix from tempClusters */
        change = 0;
        for ( i = 0; i < k; i++){
            change += calculateDistance(clusterMatrix[i], tempClusters[i], d);

            for (j = 0; j < d; j++){
                clusterMatrix[i][j] = tempClusters[i][j];
            }
        }
    




     } 
 
    for ( i = 0; i < k; i++){
        for ( j = 0; j < d; j++){
            if (j == d-1 ){
                printf("%.4f", clusterMatrix[i][j]);
            }
            else{
                printf("%.4f,", clusterMatrix[i][j]);
            }
        }
        printf("\n");
    }


    for ( i = 0; i < k; i++) {
        free(clusterMatrix[i]);
    }
    free(clusterMatrix);
    for ( i = 0; i < k; i++) {
        free(tempClusters[i]);
    }
    free(tempClusters);
    for ( i = 0; i < n; i++) {
        free(pointMatrix[i]);
    }
    free(pointMatrix);
    free(pointsInCluster);
    
    return 0;
}


